src/client/token_provisioning/mod.rs

```diff
@@ -144,6 +164,81 @@ impl<'a> FobnailClient<'a> {
+                if result.header.code == MessageClass::Response(ResponseType::Content) {
+                    *state = State::VerifyPoCertChain {
+                        chain: result.payload,
+                    }
+                } else {
+                    error!("Server gave invalid response to metadata request");
+                    state.error();
+                }
+            }
+            _ => {
+                // We already matched all possible states in handle_response()
+                // and we should never reach here
+                unreachable!()
+            }
+        }
+    }
+
+    /// Verify certchain and save it to persistent storage if verification is
+    /// successful.
+    fn verify_certchain<T>(trussed: &mut T, certmgr: &mut CertMgr, chain: &[u8]) -> Result<(), ()>
+    where
+        T: trussed::client::FilesystemClient + trussed::client::Sha256,
+    {
+        /// Minimal number of certificates required in a chain (including root).
+        const MIN_CERTS: usize = 2;
+        /// Max number of certificates allowed in a chain (including root).
+        const MAX_CERTS: usize = 3;
+
+        let chain = trussed::cbor_deserialize::<proto::PoCertChain>(chain)
+            .map_err(|e| error!("Failed to deserialize PO certchain: {}", e))?;
+
+        let num_certs = chain.certs.len();
+        if !matches!(num_certs, MIN_CERTS..=MAX_CERTS) {
+            error!(
+                "Expected between {} and {} certificates but got {}",
+                MIN_CERTS, MAX_CERTS, num_certs
+            );
+        }
+
+        let mut it = chain.certs.iter();
+        let root_raw = it.next().unwrap();
+        // Attester sends full chain, including root ca. Check only whether received
+        // root CA matches embedded CA.
+        if &root_raw[..] != CertMgr::po_root_raw() {
+            error!("Received root CA doesn't match with CA stored in firmware");
+            return Err(());
+        }
+
+        for cert in it {
+            match certmgr.load_cert_owned(cert) {
+                Ok(cert) => match certmgr.verify(trussed, &cert, crate::certmgr::VerifyMode::Po) {
+                    Ok(()) => {
+                        // Inject as volatile certificate, we will save
+                        // certificates to persistent storage only after entire
+                        // chain has been verified.
+                        certmgr.inject_volatile_cert(cert);
+                    }
+                    Err(e) => {
+                        error!("Cert verification failed: {}", e);
+                        certmgr.clear_volatile_certs();
+                        return Err(());
+                    }
+                },
+                Err(e) => {
+                    error!("Invalid certificate: {}", e);
+                    certmgr.clear_volatile_certs();
+                    return Err(());
+                }
+            }
+        }
+
+        certmgr.clear_volatile_certs();
```

Comment:
Assuming we want to check newly created certificate in 4.5, can we keep volatile certs until then?

Reply:
We can.
Certs are now kept till the end of provisioning, and are cleared from SignalStatus both on error and success.